modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -170,15 +184,72 @@ private void doWrite(ChannelHandlerContext ctx, Netty4HttpResponse readyResponse
     private void splitAndWrite(ChannelHandlerContext ctx, Netty4HttpResponse msg, ChannelPromise promise) {
         final PromiseCombiner combiner = new PromiseCombiner(ctx.executor());
         HttpResponse response = new DefaultHttpResponse(msg.protocolVersion(), msg.status(), msg.headers());
-        combiner.add(ctx.write(response));
+        combiner.add(enqueueWrite(response, ctx));
         ByteBuf content = msg.content();
         while (content.readableBytes() > SPLIT_THRESHOLD) {
-            combiner.add(ctx.write(new DefaultHttpContent(content.readRetainedSlice(SPLIT_THRESHOLD))));
+            combiner.add(enqueueWrite(new DefaultHttpContent(content.readRetainedSlice(SPLIT_THRESHOLD)), ctx));
         }
-        combiner.add(ctx.write(new DefaultLastHttpContent(content.readRetainedSlice(content.readableBytes()))));
+        combiner.add(enqueueWrite(new DefaultLastHttpContent(content.readRetainedSlice(content.readableBytes())), ctx));
         combiner.finish(promise);
     }
 
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) {
+        if (ctx.channel().isWritable()) {
+            doFlush(ctx);
+        }
+        ctx.fireChannelWritabilityChanged();
+    }
+
+    @Override
+    public void flush(ChannelHandlerContext ctx) {
+        if (doFlush(ctx) == false) {
+            ctx.flush();
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        doFlush(ctx);
+        super.channelInactive(ctx);
+    }
+
+    private boolean doFlush(ChannelHandlerContext ctx) {
```

Comment:
Can we add javadoc on the return value here?

Reply:
++ added